syntax = "proto3";

package dbx1000;
// when use brpc, this should be true;when use grpc, this should be false
option cc_generic_services = true;

enum RpcLockMode {
  O = 0;
  P = 1;
  S = 2;
  X = 3;
}

enum RpcRC {
  RCOK = 0;
  Commit = 1;
  Abort = 2;
  WAIT = 3;
  ERROR = 4;
  FINISH = 5;
  TIME_OUT = 6;
}

message LockRemoteRequest{
  int32 instance_id = 1;
  uint64 page_id = 2;
  RpcLockMode req_mode = 3;
  bytes page_buf = 4;
  uint64 count = 5;
}
message LockRemoteReply{
  RpcRC rc = 1;
  bytes page_buf = 2;
  uint64 count = 3;
}

message InstanceInitDoneRequest{
  int32 instance_id = 1;
}
message InstanceInitDoneReply{}

message GlobalLockInitDoneRequest{}
message GlobalLockInitDoneReply{
  bool init_done = 1;
}

message InvalidRequest{
  uint64 page_id = 1;
  bytes page_buf = 2;
  uint64 count = 3;
}

message InvalidReply{
  RpcRC rc = 1;
  bytes page_buf = 2;
  uint64 count = 3;
}
message GetNextTsRequest{}
message GetNextTsReply{
  uint64 ts = 1;
}

message  TestRequest{}
message TestReply{}

message ReportResultRequest{
  uint64 total_run_time = 1;
  uint64 total_latency = 2;
  uint64 total_txn_cnt = 3;
  uint64 total_time_remote_lock = 4;
  uint64 instance_run_time = 5;
  int32 instance_id = 6;
}
message ReportResultReply {}


service GlobalLockService {
  // for instance
  rpc LockRemote(LockRemoteRequest) returns (LockRemoteReply) {}
  rpc AsyncLockRemote(LockRemoteRequest) returns (LockRemoteReply) {}
  rpc InstanceInitDone(InstanceInitDoneRequest) returns (InstanceInitDoneReply) {}
  rpc AsyncInstanceInitDone(InstanceInitDoneRequest) returns (InstanceInitDoneReply) {}
  rpc GlobalLockInitDone(GlobalLockInitDoneRequest) returns (GlobalLockInitDoneReply) {}
  rpc AsyncGlobalLockInitDone(GlobalLockInitDoneRequest) returns (GlobalLockInitDoneReply) {}
  rpc GetNextTs(GetNextTsRequest) returns (GetNextTsReply) {}
  rpc ReportResult(ReportResultRequest) returns (ReportResultReply) {}

  // for lock_server
  rpc Invalid(InvalidRequest) returns (InvalidReply) {}
  rpc AsyncInvalid(InvalidRequest) returns (InvalidReply) {}
  rpc Test(TestRequest) returns (TestReply) {}
}